File: exercises/day13-Keyboard-First-Refactoring/kotlin/src/main/kotlin/sleightrip/trip/SleighTripService.kt
package sleightrip.trip

import sleightrip.exception.ElfNotLoggedInException
import sleightrip.elf.Elf
import sleightrip.elf.ElfSession

open class SleighTripService {

    fun getTripsByUser(targetElf: Elf): List<SleightTrip> {
        val sleighTrips = mutableListOf<SleightTrip>()
        val loggedElf = getLoggedUser()
        var isWorkshopFriend = false
        
        if (loggedElf != null) {
            for (friend in targetElf.getFriends()) {
                if (friend == loggedElf) {
                    isWorkshopFriend = true
                    break
                }
            }
            if (isWorkshopFriend) {
                return SleighTripDAO.findTripsByUser(targetElf)
            }
            return sleighTrips
        } else {
            throw ElfNotLoggedInException()
        }
    }

    protected open fun getLoggedUser(): Elf? {
        return ElfSession.getInstance().getLoggedUser()
    }
}

File: exercises/day13-Keyboard-First-Refactoring/scala/src/main/scala/sleightrip/trip/SleighTripService.scala
package sleightrip.trip

import sleightrip.exception.ElfNotLoggedInException
import sleightrip.elf.{Elf, ElfSession}
import scala.collection.mutable.ListBuffer

class SleighTripService {

  def getTripsByUser(targetElf: Elf): List[SleightTrip] = {
    val sleighTrips = ListBuffer[SleightTrip]()
    val loggedElf = getLoggedUser
    var isWorkshopFriend = false

    if (loggedElf != null) {
      for (friend <- targetElf.getFriends) {
        if (friend == loggedElf) {
          isWorkshopFriend = true
        }
      }
      if (isWorkshopFriend) {
        return SleighTripDAO.findTripsByUser(targetElf)
      }
      sleighTrips.toList
    } else {
      throw new ElfNotLoggedInException()
    }
  }

  protected def getLoggedUser: Elf = {
    ElfSession.getInstance().getLoggedUser
  }
}
